docs/architecture/adr-067-mempool-refactor.md

```diff
@@ -0,0 +1,215 @@
+When a proposer is ready to select transactions for the next block proposal,
+transactions are selected from this priority index by highest priority order.
+When a transaction is selected and reaped, it is removed from this index and
+from the `<sender:*Tx>` mapping.
+
+We define `Tx` as the following data structure:
+
+```go
+type Tx struct {
+  // Tx represents the raw binary transaction data.
+  Tx []byte
+
+  // Priority defines the transaction's priority as specified by the application
+  // in the ResponseCheckTx response.
+  Priority int64
+
+  // Sender defines the transaction's sender as specified by the application in
+  // the ResponseCheckTx response.
+  Sender string
+
+  // Index defines the current index in the priority queue index. Note, if
+  // multiple Tx indexes are needed, this field will be removed and each Tx
+  // index will have its own wrapped Tx type.
+  Index int
+}
+```
+
+### Eviction
+
+Upon successfully executing `CheckTx` for a new `Tx` and the mempool is currently
+full, we must check if there exists a `Tx` of lower priority that can be evicted
+to make room for the new `Tx` with higher priority and with sufficient size
+capacity left.
+
+If such a `Tx` exists, we find it by obtaining a read lock and sorting the
+priority queue index. Once sorted, we find the first `Tx` with lower priority and
+size such that the new `Tx` would fit within the mempool's size limit. We then
+remove this `Tx` from the priority queue index as well as the `<sender:*Tx>`
+mapping.
+
+This will require additional `O(n)` space and `O(n*log(n))` runtime complexity.
+
+### Gossiping
+
+When gossiping new transactions to peers, we first obtain a read lock on the
+priority queue index. If we have not gossiped the top priority `Tx` for the
+given peer, we return this `Tx` as the next transaction to gossip. Otherwise,
+we sort the priority queue index and traverse in descending priority order until
+we find the first non-sent `Tx` to the peer.
+
+In order to facilitate this, we must also keep an ephemeral cache, mapping from
+`sender` to a list of unique `Tx` identifiers (e.g. the `Tx` hash). A `Tx`
+identifier is removed from this cache once it is either committed in a block or
+gossiped to a peer. Note, if the `Tx` is committed in a block, this requires that
+we iterates over the entire cache, i.e. each peer's list.
+
+## Consequences
+
+### Positive
+
+- Transactions are allowed to be prioritized by the application.
+- Transactions are allowed to be gossiped by priority.
+- Allows future replace-by-priority functionality.
+
+### Negative
+
+- Additional bytes sent over the wire due to new fields added to `ResponseCheckTx`.
+- A transaction that passed `CheckTx` and entered the mempool can later be evicted
+  at a future point in time if a higher priority transaction entered while the
+  mempool was full.
+- Possible decreased throughput performance due to the necessity of various read
```

Comment:
It would be good if we could somehow estimate how bad this is going to be.

Reply:
I think it's gonna be tough w/o implementing a rough draft first and running throughput benchmarks of old vs new :-/

Reply:
Fair but we should be able to estimate a bit based on the data structures. We should at least have an idea of how the storage overhead from X txs and Y peers. But the actual latencies will be much harder to see without an implementation of course, though I'm a bit worried about the locking described here. But maybe it's fine for now as an alternative to the existing mempool until we have time for a more ambitious performant redesign

Reply:
I think its fine for now, and then we swap out the core sorted data structure to something with much less lock-contention in the future.